branchpro/new_posterior.py

```diff
@@ -0,0 +1,1077 @@
+
+            dLl.append(
+                (1/r_profile[_]) * np.sum(self.slice_cases[_]) - np.sum(
+                    np.multiply(self.slice_cases[_] + 1 / phi,
+                                inv_phi_r_tau_window)))
+
+            dLl_phi += np.sum(
+                digamma(self.slice_cases[_] + 1 / phi) - digamma(1 / phi) +
+                np.log(1 / phi) + 1)
+
+            dLl_phi -= np.sum(log_phi_r_tau_window) + np.sum(np.multiply(
+                self.slice_cases[_] + 1 / phi, inv_phi_r_tau_window2))
+
+        dLl_phi *= - 1 / (phi ** 2)
+
+        if self._infer_phi is True:
+            dLl.append(dLl_phi)
+
+        return dLl
+
+
+class NegBinBranchProLogPosterior(PoissonBranchProLogPosterior):
+    """NegBinBranchProLogPosterior Class:
+    Controller class for the optimisation or inference of parameters of the
+    negative binomial branching process model in a PINTS framework.
+
+    Parameters
+    ----------
+    inc_data
+        (pandas Dataframe) Dataframe of the numbers of new cases by time unit
+        (usually days).
+        Data stored in columns of with one for time and one for incidence
+        number, respectively.
+    daily_serial_interval
+        (list) Unnormalised probability distribution of that the recipient
+        first displays symptoms s days after the infector first displays
+        symptoms.
+    tau
+        (numeric) Size sliding time window over which the reproduction number
+        is estimated.
+    phi
+        (numeric) Value of the overdispersion parameter for the negative
+        binomial noise distribution.
+    alpha
+        the shape parameter of the Gamma distribution of the prior.
+    beta
+        the rate parameter of the Gamma distribution of the prior.
+    phi_shape
+        the shape parameter of the Exponential distribution of the prior of
+        the overdispersion.
+    phi_rate
+        the rate parameter of the Exponential distribution of the prior of
+        the overdispersion.
+    infer_phi
+        (boolean) Indicator value of whether the overdispersion parameter
+        for the negative binomial noise distribution is inferred or not.
+    imported_inc_data
+        (pandas Dataframe) contains numbers of imported new cases by time unit
+        (usually days).
+        Data stored in columns of with one for time and one for incidence
+        number, respectively.
+    epsilon
+        (numeric) Proportionality constant of the R number for imported cases
+        with respect to its analog for local ones.
+    time_key
+        label key given to the temporal data in the inc_data dataframe.
+    inc_key
+        label key given to the incidental data in the inc_data dataframe.
+
+    """
+    def __init__(self, inc_data, daily_serial_interval, tau, phi, alpha, beta,
+                 phi_shape, phi_rate, infer_phi=True, imported_inc_data=None,
+                 epsilon=None, time_key='Time', inc_key='Incidence Number'):
```

Comment:
If infer_phi = False, phi_shape and phi_rate need not be provided
If infer_phi = True, phi need not be provided
So these can be optional arguemnts.

Also, in the case infer_phi= True, can we have another option which is to pass a pints.LogPrior of our choice instead of shape and rate.